examples/chat/module2/rust/src/lib.rs

```diff
@@ -14,76 +14,70 @@
 // limitations under the License.
 //
 
-use chat_common::proto::chat::ReceivedMessage;
-use log::{info, warn};
+use chat_common::msg::Msg;
+use chat_common::proto::chat::Message;
+use log::info;
 use protobuf::ProtobufEnum;
 
 #[cfg(target_arch = "wasm32")]
 #[no_mangle]
 pub extern "C" fn oak_main(handle: u64) -> i32 {
-    match std::panic::catch_unwind(|| main(handle)) {
-        Ok(rc) => rc,
-        Err(_) => oak::OakStatus::ERR_INTERNAL.value(),
-    }
+    std::panic::catch_unwind(|| {
+        let room = Room::default();
+        room.event_loop(handle)
+    })
+    .unwrap_or(oak::OakStatus::ERR_INTERNAL.value())
 }
 
-pub fn main(in_handle: u64) -> i32 {
-    oak_log::init_default();
-    oak::set_panic_hook();
-    info!("per-room node started");
+#[derive(Default)]
+struct Room {
+    messages: Vec<Message>,
+    clients: Vec<oak::grpc::ChannelResponseWriter>,
+}
 
-    let mut rsp_writers: Vec<oak::grpc::ChannelResponseWriter> = Vec::with_capacity(1);
-    loop {
+impl Room {
+    fn event_loop(mut self, in_handle: u64) -> i32 {
```

Comment:
I think the new approach is a step in the right direction, but an event loop is still not a very natural or intuitive approach for most developers. In a future version we should look at also automatically generating this code so the developer only has to write the appropriate event handlers and not have to think about the mechanics of how they are triggered.